src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,363 @@
+                // or the max batch size is reached or the node is shutting down.  
+                if (dequeueTask.Status == TaskStatus.RanToCompletion)
+                {
+                    BlockPair item = dequeueTask.Result;
+
+                    // Set the dequeue task to null so it can be assigned on the next iteration.
+                    dequeueTask = null;
+
+                    batch.Add(item);
+
+                    this.currentBatchSizeBytes += item.Block.GetSerializedSize();
+
+                    saveBatch |= (item.ChainedHeader == this.chain.Tip) || (this.currentBatchSizeBytes >= BatchThresholdSizeBytes);
+                }
+                else
+                {
+                    // Will be executed in case timer ran out or node is being shut down.
+                    saveBatch = true;
+                }
+
+                if (saveBatch)
+                {
+                    if (batch.Count != 0)
+                    {
+                        await this.SaveBatchAsync(batch).ConfigureAwait(false);
+
+                        batch.Clear();
+                        this.currentBatchSizeBytes = 0;
+                    }
+
+                    timerTask = null;
+                }
+                else
+                {
+                    // Start timer if it is not started already.
+                    timerTask = timerTask ?? Task.Delay(BatchMaxSaveIntervalSeconds * 1000, this.nodeLifetime.ApplicationStopping);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Checks if repository contains reorged blocks and deletes them; saves batch on top. </summary>
+        /// <param name="batch">List of batched blocks. Cannot be empty.</param>
+        private async Task SaveBatchAsync(List<BlockPair> batch)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(batch), nameof(batch.Count), batch.Count);
+
+            List<BlockPair> batchCleared = this.GetBatchWithoutReorgedBlocks(batch);
+
+            ChainedHeader expectedStoreTip = batchCleared.First().ChainedHeader.Previous;
+
+            // Check if block repository contains reorged blocks. If it does - delete them.
+            if (expectedStoreTip.HashBlock != this.StoreTip.HashBlock)
+                await this.RemoveReorgedBlocksFromStoreAsync(expectedStoreTip).ConfigureAwait(false);
+
+            // Save batch.
+            ChainedHeader newTip = batchCleared.Last().ChainedHeader;
+
+            this.logger.LogDebug("Saving batch of {0} blocks, total size: {1} bytes.", batchCleared.Count, this.currentBatchSizeBytes);
+
+            await this.blockRepository.PutAsync(newTip.HashBlock, batchCleared.Select(b => b.Block).ToList()).ConfigureAwait(false);
+
+            this.SetStoreTip(newTip);
+
+            this.logger.LogDebug("Store tip set to '{0}'", this.StoreTip);
+            this.logger.LogTrace("(-)");
```

Comment:
in other parts of the code I've seen a new line before this logTrace but I am assuming this one doesn't have that because of the other log message and the line break above that instead?

Reply:
Fixed